packages/node/src/indexer/worker/block-dispatcher.service.ts

```diff
@@ -0,0 +1,383 @@
+      if (this.isShutdown) break;
+
+      await Promise.all(this.processQueue.putMany(blockTasks));
+    }
+
+    this.fetching = false;
+  }
+
+  get queueSize(): number {
+    return this.fetchQueue.size;
+  }
+
+  get freeSize(): number {
+    return this.fetchQueue.freeSpace;
+  }
+
+  get latestBufferedHeight(): number {
+    return this._latestBufferedHeight;
+  }
+
+  set latestBufferedHeight(height: number) {
+    this.eventEmitter.emit(IndexerEvent.BlocknumberQueueSize, {
+      value: this.queueSize,
+    });
+    this._latestBufferedHeight = height;
+  }
+}
+
+@Injectable()
+export class WorkerBlockDispatcherService
+  implements IBlockDispatcher, OnApplicationShutdown
+{
+  private workers: IndexerWorker[];
+  private numWorkers: number;
+  private getRuntimeVersion: GetRuntimeVersion;
+  private onDynamicDsCreated: (height: number) => Promise<void>;
+
+  private taskCounter = 0;
+  private isShutdown = false;
+  private queue: AutoQueue<void>;
+  private _latestBufferedHeight: number;
+
+  /**
+   * @param numWorkers. The number of worker threads to run, this is capped at number of cpus
+   * @param workerQueueSize. The number of fetched blocks queued to be processed
+   */
+  constructor(
+    numWorkers: number,
+    workerQueueSize: number,
+    private eventEmitter: EventEmitter2,
+  ) {
+    this.numWorkers = getMaxWorkers(numWorkers);
+    this.queue = new AutoQueue(this.numWorkers * workerQueueSize);
+  }
+
+  async init(
+    runtimeVersionGetter: GetRuntimeVersion,
+    onDynamicDsCreated: (height: number) => Promise<void>,
+  ): Promise<void> {
+    this.workers = await Promise.all(
+      new Array(this.numWorkers).fill(0).map(() => createIndexerWorker()),
+    );
+
+    this.getRuntimeVersion = runtimeVersionGetter;
+    this.onDynamicDsCreated = onDynamicDsCreated;
+  }
+
+  async onApplicationShutdown(): Promise<void> {
+    this.isShutdown = true;
+    // Stop processing blocks
+    this.queue.abort();
+
+    // Stop all workers
+    await Promise.all(this.workers.map((w) => w.terminate()));
+  }
+
+  enqueueBlocks(heights: number[]): void {
+    if (!heights.length) return;
+    logger.info(`Enqueing blocks ${heights[0]}...${last(heights)}`);
+
+    heights.map((height) => this.enqueueBlock(height));
+
+    this.latestBufferedHeight = last(heights);
+  }
+
+  flushQueue(height: number): void {
+    this.latestBufferedHeight = height;
+    this.queue.flush();
+  }
+
+  private enqueueBlock(height: number) {
+    if (this.isShutdown) return;
+    const workerIdx = this.getNextWorkerIndex();
+    const worker = this.workers[workerIdx];
+
+    assert(worker, `Worker ${workerIdx} not found`);
+
+    // Used to compare before and after as a way to check if queue was flushed
+    const bufferedHeight = this._latestBufferedHeight;
+    const pendingBlock = worker.fetchBlock(height);
+
+    const processBlock = async () => {
+      const start = new Date();
+      const result = await pendingBlock;
```

Comment:
If i'm not wrong, `processBlock` live in main thread, so in workthread `worker.fetchBlock()` then transfer the result to main thread. Seems something we should avoid.